From e880d3ae843ef8e41cfedb10c6a81fd1c801e471 Mon Sep 17 00:00:00 2001
From: Varun Kumar Reddy Yaparla <vyaparla@habana.ai>
Date: Thu, 23 Dec 2021 12:41:08 +0200
Subject: [PATCH] [SW-67662] gaudi2: add key validation support

This patch adds key validation support by implementing
the necessary functions in preboot to validate boot fit itb

Change-Id: I8278b10757ea619255b1a21e07daabaa11e4d1a5
---

diff --git a/arch/gaudi2/arc/Makefile b/arch/gaudi2/arc/Makefile
index 3644488..64db08b 100644
--- a/arch/gaudi2/arc/Makefile
+++ b/arch/gaudi2/arc/Makefile
@@ -7,8 +7,6 @@
 LDFLAGS += -Wl,--just-symbols $(PREBOOT_ROOT_DIR)/../arcboot-custom/exports.elf
 
 PREBOOT_SRCS_C += $(shell ls $(PREBOOT_ROOT_DIR)/lib/preboot/thermal_v2.c)
-PREBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/security.c)
 
-PPBOOT_SRCS_C += $(shell ls $(HL_DEV_ARCH)/security.c)
 image_validation: image_validation_common
 image_validation: image_validation_preboot
diff --git a/arch/gaudi2/arc/security.c b/arch/gaudi2/arc/security.c
deleted file mode 100644
index f0a0d0c..0000000
--- a/arch/gaudi2/arc/security.c
+++ /dev/null
@@ -1,88 +0,0 @@
-// SPDX-License-Identifier: GPL-2.0+
-/*
- * Copyright 2021 HabanaLabs, Ltd.
- * All Rights Reserved.
- *
- */
-
-#include <bitops.h>
-#include <preboot.h>
-#include <string.h>
-#include <uart.h>
-#include <sha512.h>
-#include <gaudi2/gaudi2_security.h>
-
-bool preboot_security_key_is_revoked(int key_idx)
-{
-	u32 key_revoke;
-
-	if (key_idx < 0 || key_idx >= SECURITY_KEY_HASH_COUNT)
-		return true;
-
-	key_revoke = sp_sram_readl(SECURITY_EFUSE_KEYS_REVOKE_OFFSET);
-	return !!(key_revoke & BIT(key_idx));
-}
-
-bool preboot_security_key_is_valid(int key_idx, const void *pub_key,
-			 int pub_key_len, const void *pub_key_ext,
-			 int pub_key_ext_len)
-{
-	u32 key_base, hash_calculated, hash_stored;
-	u8 bytes_per_dword = sizeof(u32);
-	u8 value[SHA384_SUM_LEN];
-	u8 key[KEY_SIZE];
-	u8 i;
-
-	if (key_idx < 0 || key_idx >= SECURITY_KEY_HASH_COUNT)
-		return false;
-
-	if (pub_key_len != (KEY_SIZE - KEY_EXT_SIZE)) {
-		preboot_puts_hex("Invalid pub key length: ", pub_key_len, 0);
-		return false;
-	}
-
-	if (pub_key_ext_len != KEY_EXT_SIZE) {
-		preboot_puts_hex("Invalid pub key ext length: ",
-				pub_key_ext_len, 0);
-		return false;
-	}
-
-	/* Concanate public key and public key extension */
-	memcpy(key, pub_key, pub_key_len);
-	memcpy((key + pub_key_len), pub_key_ext, pub_key_ext_len);
-
-	sha384_csum(key, pub_key_len + pub_key_ext_len, value, 0);
-
-	key_base = (key_idx * bytes_per_dword * SECURITY_DWORD_PER_KEY)
-			+ SECURITY_EFUSE_KEYS_OFFSET;
-
-	for (i = 0; i < SECURITY_DWORD_PER_KEY; i++) {
-		memcpy(&hash_calculated, value + (i * bytes_per_dword),
-			bytes_per_dword);
-		hash_stored = sp_sram_readl(key_base + (i * bytes_per_dword));
-		if (hash_calculated != hash_stored)
-			return false;
-	}
-
-	return true;
-}
-
-bool preboot_security_svn_is_valid(int sec_ver)
-{
-	u8 sec_fw_ver, sec_valid_fw_ver;
-	u32 valid_fw_ver;
-
-	sec_fw_ver = SECURITY_FW_VERSION_TO_SECURITY_BYTE(sec_ver);
-	valid_fw_ver = sp_sram_readl(SECURITY_EFUSE_FW_VERSION_OFFSET);
-	sec_valid_fw_ver = SECURITY_FW_VERSION_TO_SECURITY_BYTE(valid_fw_ver);
-
-	if (sec_fw_ver < sec_valid_fw_ver)
-		preboot_puts_hex("Security version too low: ", sec_ver, 0);
-
-	return sec_fw_ver >= sec_valid_fw_ver;
-}
-
-bool preboot_security_key_montgomery_is_valid(void)
-{
-	return false;
-}
diff --git a/arch/gaudi2/common/common.c b/arch/gaudi2/common/common.c
index 8288409..62db4c9 100644
--- a/arch/gaudi2/common/common.c
+++ b/arch/gaudi2/common/common.c
@@ -149,6 +149,11 @@
 	return sig_validation_en;
 }
 
+bool preboot_is_image_sig_supported(void)
+{
+	return preboot_img_validation_is_enabled();
+}
+
 void lkd_dev_spec_regs(struct cpu_dyn_regs *regs)
 {
 	regs->cpu_pq_base_addr_low = mmCPU_IF_PQ_BASE_ADDR_LOW;
diff --git a/arch/gaudi2/common/security.c b/arch/gaudi2/common/security.c
index b7ba9c6..cb9d03c 100644
--- a/arch/gaudi2/common/security.c
+++ b/arch/gaudi2/common/security.c
@@ -21,7 +21,9 @@
 #include <efuse.h>
 #include <psoc_security_regs.h>
 #include "sha.h"
+#include <sha256.h>
 #include <gaudi2/gaudi2_security.h>
+#include <preboot/fit.h>
 #include <psoc_arc0_mstr_if_core_hbw_regs.h>
 #include <psoc_arc0_mstr_if_core_lbw_regs.h>
 #include <psoc_arc1_mstr_if_core_hbw_regs.h>
@@ -488,6 +490,209 @@
 	preboot_security_write_otp();
 }
 
+bool preboot_security_key_montgomery_is_valid(void)
+{
+	return false;
+}
+
+bool preboot_security_key_is_revoked(int key_idx)
+{
+	const struct hl_sec_efuse *fuse;
+	int max_keys = SEC_KEY_HASH_COUNT_IN_ROM;
+
+	if (preboot_fake_security_is_enabled())
+		fuse = preboot_get_hl_sec_fake_efuse();
+	else
+		fuse = preboot_get_hl_sec_real_efuse();
+
+	if (fuse == NULL)
+		return true;
+
+	if (fuse->keys_loc_selector == SEC_KEY_HASH_LOCATION_EFUSE)
+		max_keys = SEC_KEY_HASH_COUNT_IN_EFUSE;
+
+	if (key_idx < 0 || key_idx >= max_keys)
+		return true;
+
+	if (fuse->key_revoke & BIT(key_idx))
+		return true;
+
+	return false;
+}
+
+static int preboot_security_get_minimal_svn(void)
+{
+	const struct hl_sec_efuse *fuse;
+	const u32 *svn_bits;
+	int i, svn = 0;
+
+	if (preboot_fake_security_is_enabled())
+		fuse = preboot_get_hl_sec_fake_efuse();
+	else
+		fuse = preboot_get_hl_sec_real_efuse();
+
+	if (fuse == NULL)
+		return -1;
+
+	/* Use 32bit because the data may not be 64bit aligned */
+	svn_bits = (const u32 *)fuse->sec_ver_num;
+
+	/* Count how many consecutive LSB bits out of 128 are set in 4 32bit
+	 * iterations.
+	 */
+	for (i = 0; i < SVN_EFUSE_SIZE/sizeof(u32); i++) {
+		/* No bits are set, stop counting */
+		if (!svn_bits[i])
+			break;
+		/* All bits are set, move to next iteration, bits are counted
+		 * by the iterator at the end. Move to the next iteration.
+		 */
+		if (!~svn_bits[i])
+			continue;
+
+		/* Some bits are set, count them by __builtin_ctz() the bitwise
+		 * NOT value of the SVN since __builtin_ctz() count unset bits.
+		 */
+		svn += __builtin_ctz(~svn_bits[i]);
+		break;
+	}
+	/* Add bits counted by iterator in case that all the bits are set */
+	svn += i * 32;
+
+	preboot_puts_hex("SVN on eFuse: ", svn, 0);
+	return svn;
+}
+
+bool preboot_security_svn_is_valid(int sec_ver)
+{
+	int min_svn = preboot_security_get_minimal_svn();
+
+	if (min_svn < 0)
+		return false;
+
+	if (sec_ver < min_svn || sec_ver > SVN_MAX_VERSIONS) {
+		preboot_puts_hex("Illegal security version: ", sec_ver, false);
+		return false;
+	}
+
+	return true;
+}
+
+bool preboot_security_key_is_valid(int key_idx, const void *pub_key,
+			 int pub_key_len, const void *pub_key_ext,
+			 int pub_key_ext_len)
+{
+	u8 value[SHA_SUM_LEN];
+	u32 hash_compare_len = SHA_SUM_LEN;
+	const struct hl_sec_otp *otp = preboot_security_get_otp();
+
+	if (otp == NULL)
+		return false;
+
+	if (key_idx < 0 || key_idx >= otp->num_of_keys)
+		return false;
+
+	/*
+	 * In case of real security, if security keys hash location is ROM then only the first
+	 * 256 bits are compared against the hash calculated. Since there are only 256 bits
+	 * available in PSOC registers to store keys hash
+	 */
+	if (!preboot_fake_security_is_enabled() && otp->num_of_keys == SEC_KEY_HASH_COUNT_IN_ROM)
+		hash_compare_len = SHA256_SUM_LEN;
+
+	sha_csum(pub_key, pub_key_len, value);
+
+	return !memcmp(&otp->keys[key_idx * SHA_SUM_LEN], value, hash_compare_len);
+}
+
+static struct hl_fit_sec_desc fit_sec;
+static u8 signature[FIT_SIGNATURE_SIZE];
+
+int preboot_verify_img_sig(const void *img, size_t img_size, enum fit_sec_failure *res,
+				u32 *key_idx)
+{
+	const void *fit_sec_off;
+	const void *pub_key_fit;
+
+	if (!preboot_security_is_enabled())
+		return 0;
+
+	/* hl fit files are padded so the size is 4 bytes aligned */
+	img_size = ROUND_UP4(img_size);
+
+	/* security descriptor is at the footer of the image */
+	fit_sec_off = (const void *)((u8 *)img + img_size);
+
+	/* check that the calculated memory range are legal */
+	if (!preboot_sram_is_valid_address_range((uintptr_t)fit_sec_off,
+						 sizeof(fit_sec))) {
+		preboot_puts("Invalid address range\n");
+		return 1;
+	}
+
+	memcpy(&fit_sec, fit_sec_off, FIT_SEC_DESC_SIZE);
+
+	if (fit_sec.magic_number != FIT_SEC_MAGIC) {
+		preboot_puts("Bad magic, security descriptor not found\n");
+		return 1;
+	}
+
+	if ((fit_sec.desc_version_maj != FIT_SEC_DESC_VER_MAJOR) ||
+	    (fit_sec.desc_version_min != FIT_SEC_DESC_VER_MINOR)) {
+		preboot_puts("security descriptor version mismatch\n");
+		return 1;
+	}
+
+	if (fit_sec.desc_size < FIT_SEC_DESC_SIZE) {
+		preboot_puts("security descriptor size too small\n");
+		return 1;
+	}
+
+	if (fit_sec.img_size != img_size) {
+		preboot_puts("image size mismatch with security descriptor\n");
+		return 1;
+	}
+
+	if (!preboot_security_svn_is_valid(fit_sec.sec_version)) {
+		preboot_puts("security descriptor svn not valid\n");
+		*res = FIT_SEC_SVN_INVALID;
+		return 1;
+	}
+
+	if (preboot_security_key_is_revoked(fit_sec.key_num)) {
+		preboot_puts_hex("Key revoked: ", fit_sec.key_num, false);
+		*res = FIT_SEC_KEY_REVOKED;
+		return 1;
+	}
+
+	pub_key_fit = &fit_sec.pub_key;
+
+	if (!preboot_security_key_is_valid(fit_sec.key_num, pub_key_fit,
+					   FIT_PUB_KEY_SIZE, NULL, 0)) {
+		preboot_puts("Key validation failed\n");
+		*res = FIT_SEC_SIGN_FAILED;
+		return 1;
+	}
+
+	/*
+	 * signature is calculated on both the fit image and the security
+	 * descriptor with the signature field of the descriptor zeroized.
+	 */
+	memcpy(signature, fit_sec.signature, FIT_SIGNATURE_SIZE);
+	memset(fit_sec.signature, 0, FIT_SIGNATURE_SIZE);
+
+	if (fit_verify_img_sig(img, img_size, &fit_sec, FIT_SEC_DESC_SIZE,
+			       pub_key_fit, signature, FIT_SIGNATURE_SIZE)) {
+		preboot_puts("image signature verification failed\n");
+		*res = FIT_SEC_SIGN_FAILED;
+		return 1;
+	}
+
+	*key_idx = fit_sec.key_num;
+
+	return 0;
+}
+
 #define EN_ARC_SEC_PRIV				0x3
 #define AXPROT_OVRD_SEC				0x1
 
